sympy/core/basic.py

```diff
@@ -329,7 +329,7 @@ def __ne__(self, other):
 
            but faster
         """
-        return not self.__eq__(other)
+        return not self == other
```

Comment:
The point is that ``self == other`` never returns ``NotImplemented``. If ``self.__eq__(other)`` returns ``NotImplemented``, the comparison is delegated to ``other.__eq__(self)``. If it, too, returns ``NotImplemented``, ``self == other`` evaluates to ``False``.

This way, ``a != b`` always returns the opposite of ``a == b``. If using the ``not self.__eq__(other)``pattern, this short circuits the method resolution such that both ``a != b`` and ``a == b`` may return ``False`` in some circumstances.

For other rich comparison operators (``<, >`` etc.), the method resolution is similar, and if both sides return ``NotImplemented`` a ``TypeError`` is raised.

The pull request includes tests for all of these details.